gossip/src/push_active_set.rs

```diff
@@ -0,0 +1,316 @@
+    std::collections::HashMap,
+};
+
+// Equivalent to: get_stake_bucket(Some(&u64::MAX)) + 1.
+// Verified in test_get_stake_bucket.
+const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 36;
+
+pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+
+// Keys are gossip nodes to push messages to.
+// Values are which origins the node has pruned.
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get_nodes<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
+        self.get_entry(stake).get_nodes(origin, should_force_push)
+    }
+
+    pub(crate) fn prune(
+        &self,
+        pubkey: &Pubkey,    // This node.
+        node: &Pubkey,      // Gossip node.
+        origins: &[Pubkey], // CRDS value owners.
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let stake = stakes.get(pubkey);
+        for origin in origins {
+            if origin == pubkey {
+                continue;
+            }
+            let stake = stake.min(stakes.get(origin));
+            self.get_entry(stake).prune(node, origin)
+        }
+    }
+
+    pub(crate) fn rotate<R: Rng>(
+        &mut self,
+        rng: &mut R,
+        size: usize, // Number of nodes to retain in each active-set entry.
+        cluster_size: usize,
+        nodes: &[Pubkey],
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let num_bloom_filter_items = cluster_size.max(Self::MIN_NUM_BLOOM_ITEMS);
+        let buckets: Vec<_> = nodes
+            .iter()
+            .map(|node| get_stake_bucket(stakes.get(node)))
+            .collect();
+        for (k, entry) in self.0.iter_mut().enumerate() {
+            let weights: Vec<u64> = buckets
+                .iter()
+                .map(|&bucket| {
+                    // bucket <- get_stake_bucket(min stake of {
+                    //  this node, crds value owner and gossip peer
+                    // })
+                    // weight <- (bucket + 1)^2
+                    // min stake of {...} is a proxy for how much we care about
+                    // the link, and tries to mirror similar logic on the
+                    // receiving end when pruning incoming links:
+                    // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
+                    let bucket = bucket.min(k) as u64;
```

Comment:
hmmm I'm still not sure I understand the significance of this min or `k`. Is there something significant about the order in which we're iterating the buckets in the outer loop `self.0.iter_mut().enumerate() {`?

Reply:
Added some more context in the other comment:

> On the receiving end of gossip push messages:
> https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
> we keep enough incoming links to hit a `prune_stake_threshold` but that threshold is a function of the origin's stake:
> https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
> So if the origin is unstaked or has low stake, we don't really care much about the stake of the incoming link.

We want
```
get_stake_bucket(min stake of {
   this node, crds value owner and gossip peer
})
```
as a proxy for how much we care about this link.
Here, stakes are mapped to 0..n buckets, and each bucket has its own active-set.
The k'th bucket is for origins which `min stake of {this node and the crds value owner}` maps to bucket `k`, i.e. the first 2 items in above. Taking the `min` with node's bucket results in the whole thing.

Reply:
The suggested comments were actually the other way around.
Each `PushActiveSetEntry` corresponds to a stake bucket for
```
min stake of { this node, crds value owner }
```
The entry represents set of gossip nodes to actively push to for crds values belonging to the bucket.
`nodes: &[Pubkey]` are gossip nodes to be sampled for each push active set.
I have added further comments to the code as below.
```rust
// Active set of nodes to push to are sampled from these gossip nodes,
// using sampling probabilities obtained from the stake bucket of each
// node.
let buckets: Vec<_> = nodes
    .iter()
    .map(|node| get_stake_bucket(stakes.get(node)))
    .collect();
// (k, entry) represents push active set where the stake bucket of
//     min stake of {this node, crds value owner}
// is equal to `k`. The `entry` maintains set of gossip nodes to
// actively push to for crds values belonging to this bucket.
for (k, entry) in self.0.iter_mut().enumerate() {
    let weights: Vec<u64> = buckets
        .iter()
        .map(|&bucket| {
            // bucket <- get_stake_bucket(min stake of {
            //  this node, crds value owner and gossip peer
            // })
            // weight <- (bucket + 1)^2
            // min stake of {...} is a proxy for how much we care about
            // the link, and tries to mirror similar logic on the
            // receiving end when pruning incoming links:
            // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
            let bucket = bucket.min(k) as u64;
            bucket.saturating_add(1).saturating_pow(2)
        })
        .collect();
    entry.rotate(rng, size, num_bloom_filter_items, nodes, &weights);
}
```

Reply:
> The entry represents set of gossip nodes to actively push to for crds values belonging to the bucket.
nodes: &[Pubkey] are gossip nodes to be sampled for each push active set.
I have added further comments to the code as below.

Ah thanks, that makes sense